controllers/space/space_controller.go

```diff
@@ -197,16 +196,132 @@ func (r *Reconciler) ensureNSTemplateSet(logger logr.Logger, space *toolchainv1a
+			return requeueDelay, nil
+		}
+
+		if space.Labels == nil {
+			space.Labels = map[string]string{}
+		}
+
+		// remove any outdated tier hash labels
+		for key := range space.GetLabels() {
+			if strings.HasPrefix(key, toolchainv1alpha1.LabelKeyPrefix) && strings.HasSuffix(key, "-tier-hash") {
+				delete(space.Labels, key)
+			}
+		}
+		// check parentSpace label and update it if needed
+		setParentSpaceLabel(space)
+
+		// add a tier hash label matching the current NSTemplateTier
+		hash, err := tierutil.ComputeHashForNSTemplateTier(tmplTier)
+		if err != nil {
+			err = errs.Wrap(err, "error computing hash for NSTemplateTier")
+			return norequeue, r.setStatusProvisioningFailed(logger, space, err)
+		}
+		space.Labels[tierutil.TemplateTierHashLabelKey(space.Spec.TierName)] = hash
+		if err := r.Client.Update(context.TODO(), space); err != nil {
+			return norequeue, r.setStatusProvisioningFailed(logger, space, err)
+		}
+		return norequeue, r.setStatusProvisioned(space)
+	default:
+		return norequeue, r.setStatusProvisioningFailed(logger, space, fmt.Errorf(nsTmplSetReady.Message))
+	}
+}
+
+// listSpaceBindings searches all the SpaceBindings of a given Space,
+// it will use the name of the parentSpace for the research if there is one.
+func (r *Reconciler) listSpaceBindings(space *toolchainv1alpha1.Space) (toolchainv1alpha1.SpaceBindingList, error) {
+	// list SpaceBindings bound to the space
 	spaceBindings := toolchainv1alpha1.SpaceBindingList{}
 	if err := r.Client.List(context.TODO(),
 		&spaceBindings,
 		client.InNamespace(space.Namespace),
 		client.MatchingLabels{
-			toolchainv1alpha1.SpaceBindingSpaceLabelKey: space.Name,
+			toolchainv1alpha1.SpaceBindingSpaceLabelKey: space.Name, // use current space name for the search
 		},
 	); err != nil {
-		logger.Error(err, "failed to list space bindings")
+		// return the error in case of read issues
+		return spaceBindings, err
 	}
+
+	// if there's no parentSpace set let's return the SpaceBindings from above
+	if space.Spec.ParentSpace == "" {
+		return spaceBindings, nil
+	}
+
+	// list SpaceBindings bound to the parentSpace
+	parentSpaceBindings := toolchainv1alpha1.SpaceBindingList{}
+	if err := r.Client.List(context.TODO(),
+		&parentSpaceBindings,
+		client.InNamespace(space.Namespace),
+		client.MatchingLabels{
+			toolchainv1alpha1.SpaceBindingSpaceLabelKey: space.Spec.ParentSpace, // use parentSpace name for the search
+		},
+	); err != nil {
+		// return the error and SpaceBindings from above in case read issues
+		return parentSpaceBindings, err
+	}
+
+	// if no parentSpaceBindings found
+	// let's return the spaceBindings found using the Space.Name
+	if len(parentSpaceBindings.Items) == 0 {
+		return spaceBindings, nil
+	}
+
+	// if no SpaceBindings found while searching with Space.Name
+	// let's return the spaceBindings found using the Space.Name
+	if len(spaceBindings.Items) == 0 {
+		return parentSpaceBindings, nil
+	}
```

Comment:
technically, these two parts don't have to be there as the following for-cycles will just skip it if any of the lists are empty.
But it's fine if you want to keep it there just for better reading - not objection :wink:

Reply:
I've left just the check for `if len(parentSpaceBindings.Items)` so that we avoid going into the loop if there are no parentSpaceBindings to "merge".

See: [6b9e585](https://github.com/codeready-toolchain/host-operator/pull/733/commits/6b9e5853271c52ff603d67960dbfd63ba4f10904)